techlibs/gatemate/synth_gatemate.cc

```diff
@@ -0,0 +1,396 @@
+			}
+			if (args[argidx] == "-nobram") {
+				nobram = true;
+				continue;
+			}
+			if (args[argidx] == "-noaddf") {
+				noaddf = true;
+				continue;
+			}
+			if (args[argidx] == "-nomult") {
+				nomult = true;
+				continue;
+			}
+			if (args[argidx] == "-nomx4") {
+				nomx4 = true;
+				continue;
+			}
+			if (args[argidx] == "-nomx8") {
+				nomx8 = true;
+				continue;
+			}
+			if (args[argidx] == "-dff") {
+				dff = true;
+				continue;
+			}
+			if (args[argidx] == "-retime") {
+				retime = true;
+				continue;
+			}
+			if (args[argidx] == "-noiopad") {
+				noiopad = true;
+				continue;
+			}
+			if (args[argidx] == "-noclkbuf") {
+				noclkbuf = true;
+				continue;
+			}
+			break;
+		}
+		extra_args(args, argidx, design);
+
+		if (!design->full_selection()) {
+			log_cmd_error("This command only operates on fully selected designs!\n");
+		}
+
+		log_header(design, "Executing SYNTH_GATEMATE pass.\n");
+		log_push();
+
+		run_script(design, run_from, run_to);
+
+		log_pop();
+	}
+
+	void script() override
+	{
+		if (check_label("begin"))
+		{
+			run("read_verilog -lib -specify +/gatemate/cells_sim.v +/gatemate/cells_bb.v");
+			run(stringf("hierarchy -check %s", help_mode ? "-top <top>" : top_opt.c_str()));
+		}
+
+		if (check_label("prepare"))
+		{
+			run("proc");
+			if (!noflatten) {
+				run("flatten");
+			}
+			run("tribuf -logic");
+			run("deminout");
+			run("opt_expr");
+			run("opt_clean");
+			run("check");
+			run("opt -nodffe -nosdff");
+			run("fsm");
+			run("opt");
+			run("wreduce");
+			run("peepopt");
+			run("opt_clean");
+			run("share");
+			run("techmap -map +/cmp2lut.v -D LUT_WIDTH=4");
+			run("opt_expr");
+			run("opt_clean");
+		}
+
+		if (check_label("map_mult", "(skip if '-nomult')"))
+		{
+			if (!nomult) {
+				run("techmap -map +/gatemate/mul_map.v");
+			}
+		}
+
+		if (check_label("coarse"))
+		{
+			run("alumacc");
+			run("opt");
+			run("memory -nomap");
+			run("opt_clean");
+		}
+
+		if (check_label("map_bram", "(skip if '-nobram')"))
+		{
+			if (!nobram) {
+				run("memory_bram -rules +/gatemate/brams.txt");
+				run("setundef -zero -params "
+					"t:$__CC_BRAM_CASCADE "
+					"t:$__CC_BRAM_40K_SDP t:$__CC_BRAM_20K_SDP "
+					"t:$__CC_BRAM_20K_TDP t:$__CC_BRAM_40K_TDP "
+				);
+				run("techmap -map +/gatemate/brams_map.v");
+				run("gatemate_bramopt");
+			}
+		}
+
+		if (check_label("map_ffram"))
+		{
+			run("opt -fast -mux_undef -undriven -fine");
+			run("memory_map");
+			run("opt -undriven -fine");
+		}
+
+		if (check_label("map_addf", "(skip if '-noaddf')"))
+		{
+			std::string techmap_args = "";
+			if (!noaddf) {
+				techmap_args += " -map +/gatemate/arith_map.v";
+			}
+			run("techmap -map +/techmap.v " + techmap_args);
+			run("opt -fast");
+			if (retime) {
+				run("abc -dff -D 1", "(only if -retime)");
+			}
+		}
+
+		if (check_label("map_io", "(skip if '-noiopad')"))
+		{
+			if (!noiopad) {
+				run("iopadmap -bits "
+					"-inpad $__inpad Y:I "

```

Comment:
This could be simplified by removing `$__inpad` and `$__outpad`, since `CC_IBUF` and `CC_OBUF` can be mapped by `iopadmap` directly